### PR TITLE
make safe inference of std.uuid toString() work

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

File: std/uuid.d [hunk 1]
```diff
@@ -858,38 +858,33 @@ public struct UUID
...
+                    put(sink, c);
+                else
+                    sink[i] = cast(typeof(sink[i]))c;
             }
-        }
-
-        /**
-         * Call a delegate with a string in the canonical form.
-         */
-        void toString(scope void delegate(const(char)[]) sink) const
```
Comment (WalterBright) [inline std/uuid.d]: The trouble here is that the delegate conversion removes the @safe and it's all downhill from there.

File: std/uuid.d [hunk 2]
```diff
@@ -858,38 +858,33 @@ public struct UUID
...
-                result[pos  ] = toChar!Char(hi);
+                result[pos  ] = toChar!char(hi);
                 const uint lo = (entry) & 0x0F;
-                result[pos+1] = toChar!Char(lo);
+                result[pos+1] = toChar!char(lo);
+            }
+            foreach (i, c; result)
+            {
+                static if (__traits(compiles, put(sink, c)))
+                    put(sink, c);
```
Comment (schveiguy) [inline std/uuid.d]: Shouldn't `put(sink, result)` handle all this? I thought that was the point...

Comment (WalterBright) [inline std/uuid.d]: The toString in std.uuid has messed up overloads. One of them takes a char[] argument and toString is expected to write into that. I didn't want to break that, hence the `static if`.

**Comment by andralex** [comment]
OK

**Pull request merged**